### PR TITLE
Upgrade and synchronize dygraphs, fix legend fall-off

### PR DESCRIPTION
**Pull request by jaredscheib**
- [x] CHANGELOG.md updated
  - [x] Rebased/mergable
  - [x] Tests pass
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

Connect #1469 #1480 #1473 

### The problem
- Legend would fall off the bottom of the page
- We wanted to synchronize hover line (vertical crosshair) and time range across graphs

### The Solution
- Upgrade to dygraphs2.0.0, which went through 4 release candidates and is stable. Upgrading makes synchronization easier, as that release includes at least 5 fixes explicitly in support of synchronization.
- Implement synchronized vertical crosshair (aka hover line) and time range across all graphs:

- Fix legend fall off bug:
screen shot 2017-05-19 at 5 53 10 pm

- Ensure that only graphs retain their expect x-axis range and only show the hover line when there's actual overlap
synchro-with-time-group

### Notes
- Future legend issues may want to track this issue: https://github.com/danvk/dygraphs/issues/846

### CODE REVIEW

File: ui/src/shared/components/Dygraph.js [hunk 1]
```diff
@@ -106,55 +103,66 @@ export default React.createClass({
...
+
         // Enforcing max & min legend offsets
         if (trueGraphX < legendWidth / 2) {
           legendLeft = legendWidth / 2
         } else if (trueGraphX > legendMaxLeft) {
           legendLeft = legendMaxLeft
         }
 
+        // Disallow screen overflow of legend
+        const legendBottomClipped = graphBottom + legendHeight > screenHeight
```
Comment (cryptoquick) [inline ui/src/shared/components/Dygraph.js]: Are we clear on the order of operations here?

File: ui/src/dashboards/components/Dashboard.js [hunk 2]
```diff
@@ -23,7 +24,7 @@ const Dashboard = ({
   onSelectTemplate,
   showTemplateControlBar,
 }) => {
-  if (dashboard.id === 0) {
+  if (!dashboard) {
```
Comment (lukevmorris) [inline ui/src/dashboards/components/Dashboard.js]: Looks like this behavior change isn't tied to anything in this PR -- was this just old code?

Comment (jaredscheib) [inline ui/src/dashboards/components/Dashboard.js]: Not sure. Looks like this was a @121watts edit

File: ui/src/shared/components/Dygraph.js [hunk 3]
```diff
@@ -200,17 +211,61 @@ export default React.createClass({
...
+    // if (this.lastMouseMoveEvent) {
+    //   dygraph.mouseMove_(this.lastMouseMoveEvent)
+    // }
 
     dygraph.resize()
-  },
+  }
+
+  sync() {
+    if (this.props.synchronizer && !this.state.synced) {
```
Comment (jaredscheib) [inline ui/src/shared/components/Dygraph.js]: Maybe this was written to protect future graphs, including `<MiniGraph>` (not used in the app currently) from being synced automatically. I'll change up the logic a bit.

File: ui/src/shared/components/Dygraph.js (hunk 1 above)
Comment (jaredscheib) [inline ui/src/shared/components/Dygraph.js]: Yep, works as intended -- addition is happening prior to comparison. Prettier actually removed parens if I try to add them 😈

**Pull request merged**